File: Labs/experiment2/E_four_number_swap_sort.cpp
#include <iostream>
using namespace std;

int lst[1000005], cir[1000005];
bool vis[1000005];

int main() {
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++) cin >> lst[i];
    for (int i = 1; i <= n; i++) {
        if (lst[i] == i || vis[i]) continue;
        int res = 0, tmp = i;
        while (!vis[tmp]) {
            res++;
            vis[tmp] = 1;
            tmp = lst[tmp];
        }
        cir[res]++;
    }
    long long ans = 0;
    for (int i = 5; i <= n; i++) {
        ans += i * cir[i] / 3;
        if (i % 3 == 2) cir[2] += cir[i];
    }
    ans += (cir[2] + 1) / 2 + cir[3] + cir[4];
    cout << ans << endl;
    return 0;
}

File: Labs/experiment2/E_four_number_swap_sort.py
n = int(input())
lst = [0] + list(map(int, input().split()))
vis = [False] * (n + 1)
cir = [0] * (n + 1)

for i in range(1, n + 1):
    if lst[i] == i or vis[i]:
        continue
    res = 0
    tmp = i
    while not vis[tmp]:
        res += 1
        vis[tmp] = True
        tmp = lst[tmp]
    cir[res] += 1

ans = 0
for i in range(5, n + 1):
    ans += i * cir[i] // 3
    if i % 3 == 2:
        cir[2] += cir[i]
ans += (cir[2] + 1) // 2 + cir[3] + cir[4]
print(ans)
